src/time/Storage.cpp

```diff
@@ -124,6 +139,73 @@ time::Sample Storage::computeExtrapolation()
   PRECICE_UNREACHABLE("Invalid _extrapolationOrder")
 }
 
+// helper function to compute x(t) from given data (x0,t0), (x1,t1), ..., (xn,tn) via B-spline interpolation (implemented using Eigen).
+Eigen::VectorXd Storage::bSplineInterpolationAt(double t, Eigen::VectorXd ts, Eigen::MatrixXd xs, int splineDegree)
+{
+  // organize data in columns. Each column represents one sample in time.
+  PRECICE_ASSERT(xs.cols() == ts.size());
+  const int ndofs = xs.rows(); // number of dofs. Each dof needs it's own interpolant.
+
+  Eigen::VectorXd interpolated(ndofs);
+
+  const int splineDimension = 1;
+
+  for (int i = 0; i < ndofs; i++) {
+    auto spline     = Eigen::SplineFitting<Eigen::Spline<double, splineDimension>>::Interpolate(xs.row(i), splineDegree, ts);
+    interpolated[i] = spline(t)[0]; // get component of spline associated with xs.row(i)
+  }
+
+  return interpolated;
+}
+
+Eigen::VectorXd Storage::sampleAt(double normalizedDt)
+{
+  const int usedOrder = computeUsedOrder(_interpolationOrder, nTimes());
+
+  PRECICE_ASSERT(math::equals(this->maxStoredNormalizedDt(), time::Storage::WINDOW_END), this->maxStoredNormalizedDt()); // sampling is only allowed, if a window is complete.
+
+  if (_interpolationOrder == 0) {
+    return this->getValuesAtOrAfter(normalizedDt);
+  }
+
+  PRECICE_ASSERT(usedOrder >= 1);
+
+  auto data = getTimesAndValues();
+  return bSplineInterpolationAt(normalizedDt, data.first, data.second, usedOrder);
+}
+
+Eigen::MatrixXd Storage::sampleGradientsAt(double normalizedDt)
+{
+  const int usedOrder = computeUsedOrder(_interpolationOrder, nTimes());
+
+  PRECICE_ASSERT(math::equals(this->maxStoredNormalizedDt(), time::Storage::WINDOW_END), this->maxStoredNormalizedDt()); // sampling is only allowed, if a window is complete.
+
+  if (_interpolationOrder == 0) {
+    return this->getGradientsAtOrAfter(normalizedDt);
+  }
+
+  PRECICE_WARN("You specified interpolation degree of {}, but only degree 0 is supported for gradient interpolation"); // @todo implement this like for sampleAt
+  return this->getGradientsAtOrAfter(normalizedDt);
```

Comment:
Here, we also should offer higher-order interpolation for gradients. We should also add a corresponding unit test, because this feature is rarely used.